File: driver/js/packages/hippy-web-renderer/src/component/refresh-wrapper.ts
/*
 * Tencent is pleased to support the open source community by making
 * Hippy available.
 *
 * Copyright (C) 2022 THL A29  Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { NodeProps, HippyBaseView, InnerNodeTag } from '../types';

import { setElementStyle } from '../common';
import { HippyWebView } from './hippy-web-view';

const BounceBackTime = 200;
const BounceBackEasingFunction = 'cubic-bezier(0.645, 0.045, 0.355, 1)';
const PullOverThreshold = 55;
const PullOverStage = [
  { threshold: [0, 50], damp: 1 },
  { threshold: [50, 100], damp: 0.6 },
  { threshold: [100, Number.MAX_SAFE_INTEGER], damp: 0.2 },
];

export class RefreshWrapper extends HippyWebView<HTMLDivElement> {
  private pullRefresh!: PullRefresh;
  public constructor(context, id, pId) {
    super(context, id, pId);
    this.tagName = InnerNodeTag.REFRESH;
    this.dom = document.createElement('div');
  }

  public defaultStyle() {
    return { display: 'flex', flexDirection: 'column', overflow: ' hidden', boxSizing: 'border-box', position: 'relative' };
  }

  public get bounceTime() {
    return this.props[NodeProps.BOUNCE_TIME];
  }

  public set bounceTime(value: number) {
    this.props[NodeProps.BOUNCE_TIME] = value;
  }

  public onRefresh() {
    this.props[NodeProps.ON_REFRESH]
    && this.context.sendUiEvent(this.id, NodeProps.ON_REFRESH, null);
  }

  public refreshComplected() {
    this.pullRefresh?.finish();
  }

  public startRefresh() {
    this.pullRefresh?.start();
  }

  public async beforeChildMount(child: HippyBaseView, childPosition: number) {
    await super.beforeChildMount(child, childPosition);
    if (child.tagName === InnerNodeTag.LIST) {
      setElementStyle(child.dom!, { position: 'relative', zIndex: 2 });
      this.pullRefresh = new PullRefresh(
        child.dom!, this.dom!.childNodes[0] as HTMLElement,
        this.handlePull.bind(this),
      );
      this.pullRefresh.init();
    }
  }

  public async beforeRemove(): Promise<void> {
    await super.beforeRemove();
    this.pullRefresh?.destroy();
  }

  private handlePull() {
    if (this.props[NodeProps.ON_REFRESH]) {
      this.onRefresh();
      return;
    }
    setTimeout(() => {
      this.pullRefresh?.finish();
    }, 500);
  }
}
export class RefreshWrapperItemView extends HippyWebView<HTMLDivElement> {
  public constructor(context, id, pId) {
    super(context, id, pId);
    this.tagName = InnerNodeTag.REFRESH_ITEM;
    this.dom = document.createElement('div');
  }
}
class PullRefresh {
  private refreshContent: HTMLElement;
  private scrollContent: HTMLElement;
  private contentStyleCache: any;
  private lastPosition = 0;
  private lastRefreshTime: number = Date.now();
  private startPosition = 0;
  private touchMove = 0;
  private moveLengthRecord = 0;
  private refreshStatus = false;
  private refreshHeadHeight = 0;
  private handleCallBack: () => void;
  private isPullFirst = false;
  public constructor(scrollContent: HTMLElement, refreshContent: HTMLElement, pullCallback: () => void) {
    this.refreshContent = refreshContent;
    this.scrollContent = scrollContent;
    this.contentStyleCache = scrollContent.style;
    this.refreshStatus = false;
    this.handleCallBack = pullCallback;
    this.handleTouchStart = this.handleTouchStart.bind(this);
    this.handlerTouchMove = this.handlerTouchMove.bind(this);
    this.handlerTouchEnd = this.handlerTouchEnd.bind(this);
  }

  public get overScrollThreshold() {
    return this.refreshHeadHeight > PullOverThreshold ? this.refreshHeadHeight : PullOverThreshold;
  }

  public get endYOffset() {
    return 0 - this.refreshContent?.clientHeight;
  }

  public init() {
    this.contentStyleCache = this.scrollContent.style;
    this.scrollContent.addEventListener('touchstart', this.handleTouchStart);
    this.scrollContent.addEventListener('touchmove', this.handlerTouchMove);
    this.scrollContent.addEventListener('touchend', this.handlerTouchEnd);
    this.resetRefreshContentTop();
  }

  public resetRefreshContentTop() {
    setElementsStyle([this.refreshContent], { transform:
        buildTranslate(0, `${(-this.refreshContent.clientHeight) ?? 0}px`) });
  }

  public finish() {
    setElementsStyle([this.refreshContent], {
      transform: buildTranslate(0, `${this.endYOffset}px`),
      transition: buildTransition(
        'transform',
        `${BounceBackTime / 1000}`,
        BounceBackEasingFunction,
      ),
    });
    setElementsStyle([this.scrollContent], {
      transform: buildTranslate(0, 0),
      transition: buildTransition(
        'transform',
        `${BounceBackTime / 1000}`,
        BounceBackEasingFunction,
      ),
    });
  }

  public start() {
    if (!this.isPullFirst) this.handlePull(0);
    setElementsStyle([this.scrollContent, this.refreshContent], {
      transform: buildTranslate(0, `${this.overScrollThreshold}px`),
      transition: buildTransition(
        'transform',
        `${BounceBackTime / 1000}`,
        BounceBackEasingFunction,
      ),
    });
    this.moveLengthRecord = this.overScrollThreshold;
    this.refreshStatus = true;
    setTimeout(() => {
      this.handleCallBack?.();
    }, BounceBackTime);
  }

  public destroy() {
    this.scrollContent.removeEventListener('touchstart', this.handleTouchStart);
    this.scrollContent.removeEventListener('touchmove', this.handlerTouchMove);
    this.scrollContent.removeEventListener('touchend', this.handlerTouchEnd);
  }

  private handlePull(moveLen) {
    const realMove = simulationBounce(moveLen);
    if (this.moveLengthRecord >= 0) {
      this.isPullFirst = true;
      setElementStyle(this.scrollContent, { transform: buildTranslate(0, `${realMove}px`) });
      if (
        this.refreshContent
        && this.refreshContent.tagName === InnerNodeTag.REFRESH_ITEM
        && (this.refreshContent.style.top === undefined
          || this.refreshContent.style.top === '')
      ) {
        this.refreshHeadHeight = this.refreshContent.clientHeight;
        setElementStyle(this.refreshContent, { top: -this.refreshContent.clientHeight });
      }
      setElementStyle(this.refreshContent, { transform:
          buildTranslate(0, `${(realMove - this.refreshContent.clientHeight) ?? 0}px`) });
      this.refreshStatus = this.moveLengthRecord >= this.overScrollThreshold;
    }
    return realMove;
  }

  private handleTouchStart(e: TouchEvent) {
    this.startPosition = e.touches[0].pageY;
    setElementsStyle([this.scrollContent, this.refreshContent], {
      transition: buildTransition('transform', 0),
    });
  }

  private handlerTouchMove(e: TouchEvent) {
    if (this.scrollContent && this.scrollContent.scrollTop > 0) {
      return;
    }
    if (!this.lastPosition) {
      this.lastPosition = e.touches[0].pageY;
    }
    if (
      Math.abs(e.touches[0].pageY - this.lastPosition) < 1
      || Date.now() - this.lastRefreshTime < 16
    ) {
      return;
    }
    this.lastRefreshTime = Date.now();
    this.touchMove += e.touches[0].pageY - this.lastPosition;
    this.lastPosition = e.touches[0].pageY;
    this.moveLengthRecord = this.handlePull(this.touchMove);
  }

  private handlerTouchEnd() {
    if (this.moveLengthRecord > 0) {
      setElementsStyle([this.scrollContent, this.refreshContent], {
        transition: buildTransition(
          'transform',
          `${BounceBackTime / 1000}`,
          BounceBackEasingFunction,
        ),
      });
      if (this.refreshStatus) {
        this.handleCallBack?.();
        setElementsStyle([this.scrollContent], {
          transform: buildTranslate(0, `${this.overScrollThreshold}px`),
        });
        setElementsStyle([this.refreshContent], {
          transform: buildTranslate(0, `${(this.overScrollThreshold - this.refreshContent.clientHeight) ?? 0}px`),
        });
      } else {
        this.finish();
        setTimeout(() => {
          this.moveLengthRecord = 0;
          this.touchMove = 0;
        }, BounceBackTime);
        return;
      }
    }
    this.moveLengthRecord = 0;
    this.touchMove = 0;
    this.lastPosition = 0;
  }
}

function simulationBounce(scrollLength: number) {
  let needMoveLength = 0;
  for (let i = 0; i < PullOverStage.length; i++) {
    const stage = PullOverStage[i];
    if (scrollLength < stage.threshold[1] && scrollLength >= stage.threshold[0]) {
      needMoveLength += (scrollLength - stage.threshold[0]) * stage.damp;
    }
    if (scrollLength >= stage.threshold[1]) {
      needMoveLength += (stage.threshold[1] - stage.threshold[0]) * stage.damp;
    }
  }
  return needMoveLength;
}
function buildTranslate(x: number | string, y: number | string) {
  return `translate(${x}, ${y})`;
}
function buildTransition(propsName: string, time: string | number, timeFunction = 'ease-out') {
  return `${propsName} ${time}s ${timeFunction}`;
}
function setElementsStyle(elements: Array<HTMLElement>, style: any) {
  elements.forEach((item) => {
    setElementStyle(item, style);
  });
}

File: renderer/native/android/src/main/java/com/tencent/mtt/hippy/views/refresh/RefreshWrapper.java
/* Tencent is pleased to support the open source community by making Hippy available.
 * Copyright (C) 2018 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.mtt.hippy.views.refresh;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.AccelerateInterpolator;

import com.tencent.mtt.hippy.utils.PixelUtil;
import com.tencent.mtt.hippy.views.common.ClipChildrenView;
import com.tencent.mtt.hippy.views.hippylist.HippyRecyclerViewWrapper;
import com.tencent.mtt.hippy.views.view.HippyViewGroup;
import com.tencent.renderer.utils.EventUtils;
import java.util.HashMap;
import java.util.Map;

@SuppressWarnings({"deprecation", "unused"})
public class RefreshWrapper extends HippyViewGroup implements ClipChildrenView {

    RefreshWrapperItemView mRefreshWrapperItemView;
    View mContentView;

    float mTansY = -1;
    float mStartTransY = 0;
    float mStartY = 0;
    float mStartDownY = -1;
    float mStartX = 0;

    RefreshState mState = RefreshState.Init;

    int mBounceTime = 300;
    protected boolean mScrollEventEnable = true;
    protected int mScrollEventThrottle = 400;  // 400ms最多回调一次
    private long mLastScrollEventTimeStamp = -1;

    public void refreshComplected() {
        bounceToHead(0);
        mState = RefreshState.Init;
    }

    public void setTime(int time) {
        this.mBounceTime = time;
    }

    enum RefreshState {
        Init,
        Loading,
    }

    public RefreshWrapper(Context context) {
        super(context);
        setClipChildren(true);
    }

    void setSTranslationY(float y) {
        if (mRefreshWrapperItemView != null) {
            mRefreshWrapperItemView.setTranslationY(y > 0 ? y : 0);
        }
        if (mContentView != null) {
            mContentView.setTranslationY(y > 0 ? y : 0);
        }
    }

    public void setOnScrollEventEnable(boolean enable) {
        mScrollEventEnable = enable;
    }

    public void setScrollEventThrottle(int scrollEventThrottle) {
        mScrollEventThrottle = scrollEventThrottle;
    }

    void bounceToHead(float toTransY) {
        if (mContentView != null && mRefreshWrapperItemView != null) {
            Animator contentAnimator = ObjectAnimator
                    .ofFloat(mContentView, "TranslationY", mContentView.getTranslationY(),
                            toTransY);
            contentAnimator.setDuration(mBounceTime);
            contentAnimator.setInterpolator(new AccelerateInterpolator());
            Animator wrapperAnimator = ObjectAnimator
                    .ofFloat(mRefreshWrapperItemView, "TranslationY",
                            mRefreshWrapperItemView.getTranslationY(),
                            toTransY);
            wrapperAnimator.setInterpolator(new AccelerateInterpolator());
            wrapperAnimator.setDuration(mBounceTime);
            contentAnimator.start();
            wrapperAnimator.start();
        }
    }


    float getCompactScrollY() {
        if (mContentView instanceof HippyRecyclerViewWrapper) {
            return ((HippyRecyclerViewWrapper) mContentView).getRecyclerView().getContentOffsetY();
        }
        return mContentView.getScrollY();
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        float nowMoveY = event.getRawY();
        float nowMoveX = event.getRawX();
        if (mContentView != null && mRefreshWrapperItemView != null) {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    mStartY = event.getRawY();
                    mStartX = event.getRawX(); //记录首次DOWN的位置
                    mStartTransY = mRefreshWrapperItemView.getTranslationY();

                    break;
                case MotionEvent.ACTION_MOVE:

                    if (nowMoveY - mStartY > 0) {
                        float scrollY = getCompactScrollY();

                        if (scrollY == 0) {
                            if (mStartDownY == -1) {
                                mStartDownY = nowMoveY;
                            } else {
                                mTansY = (nowMoveY - mStartDownY) / 3;
                                setSTranslationY(mTansY + mStartTransY);
                                //把y的偏移值返回给js侧,这里把偏移值转换为负数，方便前端理解
                                sendOnScrollEvent(-mTansY);
                            }

                            //如果垂直滚动得比水平滚动得多，就认为当前用户手势是一个上下拉，RefreshWrapper就把事件吃掉，RefreshWrapper里面的孩子就收不到事件了
                            //否则RefreshWrapper里面的孩子能收到事件，ViewPager还能继续处理左右滚动事件。
                            if (Math.abs(nowMoveX - mStartX) < Math.abs((nowMoveY - mStartY))) {
                                return true;
                            }
                        }
                    } else {
                        if (mState == RefreshState.Loading) {
                            float dis = nowMoveY - mStartY;
                            if (mRefreshWrapperItemView.getTranslationY() > 0) {
                                setSTranslationY(mStartTransY + dis);
                                //如果垂直滚动得比水平滚动得多，就认为当前用户手势是一个上下拉，RefreshWrapper就把事件吃掉，RefreshWrapper里面的孩子就收不到事件了
                                //否则RefreshWrapper里面的孩子能收到事件，ViewPager还能继续处理左右滚动事件。
                                if (Math.abs(nowMoveX - mStartX) < Math.abs(dis)) {
                                    return true;
                                }
                            }
                        }
                    }
                    break;
                case MotionEvent.ACTION_UP:
                    if (mState == RefreshState.Init) {
                        if (mTansY < mRefreshWrapperItemView.getHeight()
                                && mRefreshWrapperItemView.getTranslationY() > 0) {
                            bounceToHead(0);
                            //如果是水平滚动，不发送cancle事件。
                            if (Math.abs(nowMoveX - mStartX) < Math.abs((nowMoveY - mStartY))) {
                                sendCancelEvent(event);
                            }
                        } else if (mTansY > mRefreshWrapperItemView.getHeight()) {
                            startRefresh();
                            //如果是水平滚动，不发送cancle事件。
                            if (Math.abs(nowMoveX - mStartX) < Math.abs((nowMoveY - mStartY))) {
                                sendCancelEvent(event);
                            }
                        }
                    } else if (mState == RefreshState.Loading) {
                        if (mRefreshWrapperItemView.getTranslationY() > mRefreshWrapperItemView
                                .getHeight()) {
                            startRefresh();
                            //如果是水平滚动，不发送cancle事件。
                            if (Math.abs(nowMoveX - mStartX) < Math.abs((nowMoveY - mStartY))) {
                                sendCancelEvent(event);
                            }
                        }
                    }
                    mStartDownY = -1;
                    break;
            }
        }
        return super.dispatchTouchEvent(event);
    }

    public void sendCancelEvent(MotionEvent event) {
        MotionEvent motionEvent = MotionEvent.obtain(event);
        motionEvent.setAction(MotionEvent.ACTION_CANCEL);
        mContentView.dispatchTouchEvent(motionEvent);
    }


    public void startRefresh() {
        mTansY = -1;
        bounceToHead(mRefreshWrapperItemView.getHeight());
        mState = RefreshState.Loading;
        EventUtils.sendComponentEvent(this, EventUtils.EVENT_REFRESH_WRAPPER_REFRESH, null);
    }

    public void sendOnScrollEvent(float y) {
        if (mScrollEventEnable) {
            long currTime = System.currentTimeMillis();
            if (currTime - mLastScrollEventTimeStamp < mScrollEventThrottle) {
                return;
            }
            EventUtils.sendComponentEvent(this, EventUtils.EVENT_REFRESH_WRAPPER_SCROLL, generateScrollEvent(y));
            mLastScrollEventTimeStamp = currTime;
        }
    }

    private Map<String, Object> generateScrollEvent(float y) {
        Map<String, Object> contentOffset = new HashMap<>();
        contentOffset.put("x", PixelUtil.px2dp(0));
        contentOffset.put("y", PixelUtil.px2dp(y));
        Map<String, Object> event = new HashMap<>();
        event.put("contentOffset", contentOffset);
        return event;
    }

    @Override
    public void addView(View child, int index) {
        if (child instanceof RefreshWrapperItemView) {
            mRefreshWrapperItemView = (RefreshWrapperItemView) child;
        } else {
            mContentView = child;
        }
        super.addView(child, index);
    }

}
